File: 0105-construct-binary-tree-from-preorder-and-inorder-traversal/0105-construct-binary-tree-from-preorder-and-inorder-traversal.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    int idx=0;
    private TreeNode helper(int[] preorder, int s, int e, HashMap<Integer, Integer> map){
        if(s>e){
            return null;
        }
        int val=preorder[idx++];
        TreeNode root=new TreeNode(val);
        int index=map.get(val);
        root.left=helper(preorder,s,index-1,map);
        root.right=helper(preorder,index+1,e,map);

        return root;
    }
    public TreeNode buildTree(int[] preorder, int[] inorder) {
        HashMap<Integer,Integer> map=new HashMap<>();
        int n=preorder.length;
        for(int i=0;i<n;i++){
            map.put(inorder[i],i);
        }

        return helper(preorder,0,n-1,map);
    }
}

File: 0105-construct-binary-tree-from-preorder-and-inorder-traversal/0105-construct-binary-tree-from-preorder-and-inorder-traversal.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
   int getindex(vector<int>& preorder, vector<int>& inorder, int &index){
    int ind=-1;
    for(int i=0;i<inorder.size();i++){
        if(inorder[i]==preorder[index]){
            return i;
        }
    }
    return ind;
   }
    TreeNode *f(vector<int>& preorder, vector<int>& inorder,int s, int e, int &index){
        if(index>=preorder.size()){
            return NULL;
        }
        if(s>e){
            return NULL;
        }
        
        TreeNode *root=new TreeNode(preorder[index]);
        int mid=getindex(preorder,inorder,index);
        index++;
        root->left=f(preorder,inorder,s,mid-1,index);
        root->right=f(preorder,inorder,mid+1,e,index);
        return root;
    }
    TreeNode* buildTree(vector<int>& preorder, vector<int>& inorder) {
        int index=0;
        int n=inorder.size()-1;
        return f(preorder,inorder,0,n,index);
    }
};